storage: avoid resurrecting dead replicas

petermattis opened the pull request:
Introduce Replica.mu.minReplicaID which enforces the tombstone invariant
that we don't accept messages for previous Replica incarnations.

Make TestRaftRemoveRace more aggressive.

Fixes #9037

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10690)
<!-- Reviewable:end -->

petermattis committed:
storage: avoid resurrecting dead replicas

Introduce Replica.mu.minReplicaID which enforces the tombstone invariant
that we don't accept messages for previous Replica incarnations.

Make TestRaftRemoveRace more aggressive.

Fixes #9037

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2001 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWZcoPlUCgAOo-tvY8Z:-KWZcoPmUap6xV9nfqlm:b-phfigm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/client_raft_test.go#L2001)):*
> ```Go
> 
> 	const rangeID = roachpb.RangeID(1)
> 	mtc.replicateRange(rangeID, 1, 2, 3, 4, 5, 6, 7, 8, 9)
> ```

did this increase the probability of triggering #9037?

---

*[pkg/storage/replica_test.go, line 6040 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWZcxpWLDOsxCSNxUv0:-KWZcxpWLDOsxCSNxUv1:b-wx1uqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/replica_test.go#L6040)):*
> ```Go
> 		{2, 0, 1, "replicaID cannot move backwards"},
> 	}
> 	for i, c := range testCases {
> ```

nit: better to give these guys names rather than numeric indices.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-:-KWZd3RBjzG10pZiRubF:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2001 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWZcoPlUCgAOo-tvY8Z:-KWZlo5drEPdD2kzoJ-J:b58qbs9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/client_raft_test.go#L2001)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> did this increase the probability of triggering #9037?
</details>
Heh, I thought that was necessary per your debugging on #9037. Let me see if the problem reproduces without this change.

---

*[pkg/storage/replica_test.go, line 6040 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWZcxpWLDOsxCSNxUv0:-KWZmH51fDtQCu9S8Wlu:b65c65c) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/replica_test.go#L6040)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: better to give these guys names rather than numeric indices.
</details>
I just tried to do that, and yet coming with names is onerous. The first two cases are both success, should they be `Success1` and `Success2`. The next case could be something like `NewIDLessThanMinID`. The 4th case would be `NewIDLessThanMinIDExistingIDNonZero`. This doesn't seem to be adding anything to these tests and will place additional burden on adding new test cases in the future.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2001 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWZcoPlUCgAOo-tvY8Z:-KWZnraUOm2D8O0b895L:b-emdr6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/client_raft_test.go#L2001)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Heh, I thought that was necessary per your debugging on #9037. Let me see if the problem reproduces without this change.
</details>
Yes, this absolutely increases the probability. Without this change to`TestRaftRemoveRace` the test survives 5m of stress (prior to the fix in this PR). With this change it dies in less than a minute. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2001 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWZcoPlUCgAOo-tvY8Z:-KWZoLi4Ccl3q43N6rjK:b732zij) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/client_raft_test.go#L2001)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yes, this absolutely increases the probability. Without this change to`TestRaftRemoveRace` the test survives 5m of stress (prior to the fix in this PR). With this change it dies in less than a minute. 
</details>
Excellent, thanks for checking.

---

*[pkg/storage/replica_test.go, line 6040 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWZcxpWLDOsxCSNxUv0:-KWZoQB6EPMgXWk3Oykh:b53wlri) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/replica_test.go#L6040)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I just tried to do that, and yet coming with names is onerous. The first two cases are both success, should they be `Success1` and `Success2`. The next case could be something like `NewIDLessThanMinID`. The 4th case would be `NewIDLessThanMinIDExistingIDNonZero`. This doesn't seem to be adding anything to these tests and will place additional burden on adding new test cases in the future.
</details>
merp, fine. just sayin', the names can include spaces.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM, but is this really enough? I thought what we were discovering in #9037 was that we needed to do something (write a snapshot or set minReplicaID)  when we apply a snapshot over an existing replica without GC. 

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 1991 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWabROe4lx3F2nCAMN7:-KWabROfpmpyU9MBmzt2:b-y55716) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/client_raft_test.go#L1991)):*
> ```Go
> // TestRaftRemoveRace adds and removes a replica repeatedly in an
> // attempt to reproduce a race
> // (https://github.com/cockroachdb/cockroach/issues/1911). Note that
> ```

Add #9037 to this comment. Now that we use the `stress` tool so much we probably don't need to talk as much about how many iterations are needed to reproduce a problem. 

---

*[pkg/storage/client_raft_test.go, line 2001 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWabDpPWDJ260ndw09I:-KWabDpPWDJ260ndw09J:b62nsbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/client_raft_test.go#L2001)):*
> ```Go
> 
> 	const rangeID = roachpb.RangeID(1)
> 	mtc.replicateRange(rangeID, 1, 2, 3, 4, 5, 6, 7, 8, 9)
> ```

This needs a comment about how the extra heartbeat traffic (right?) from the additional replicas makes races more likely to occur. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-:-KWacZ20o5olT74BWZgD:b-wkxk0a)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I added setting of `Replica.mu.minReplicaID` to `setReplicaIDLocked` when a preemptive snapshot is being performed. I also write the tombstone key at this point, though I'm not sure that is necessary. PTAL.

---

Review status: 2 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_raft_test.go, line 1991 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWabROe4lx3F2nCAMN7:-KWcGP6WK8JYxiWixoP3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/client_raft_test.go#L1991)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Add #9037 to this comment. Now that we use the `stress` tool so much we probably don't need to talk as much about how many iterations are needed to reproduce a problem. 
</details>
Done.

---

*[pkg/storage/client_raft_test.go, line 2001 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWabDpPWDJ260ndw09I:-KWcGSdLqRt2MBIWySqM:b-km0ujn) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/client_raft_test.go#L2001)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This needs a comment about how the extra heartbeat traffic (right?) from the additional replicas makes races more likely to occur. 
</details>
Added a comment, though I'm not sure it is the heartbeat traffic. Perhaps the extra replicas cause concurrent down-replication. But I wasn't able to reproduce the problem with 4 replicas.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-:-KWcGfwAYteNdXkXq0vo:b-lzgpbf)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
We need a test where setting minReplicaID and writing the tombstone during preemptive snapshots matters.

---

Review status: 2 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 744 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcKamm2h3OHG8mCOX4:-KWcKamm2h3OHG8mCOX5:bgz7yde) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/replica.go#L744)):*
> ```Go
> 		r.mu.minReplicaID = r.mu.state.Desc.NextReplicaID
> 		ctx := r.AnnotateCtx(context.TODO())
> 		// TODO(peter): Is writing the tombstone necessary? If this replica is
> ```

I believe persisting the tombstone to disk is necessary. Otherwise we could still get messages for the old replica ID, which would set the replica ID of our replica and cause the same problems.

---

*[pkg/storage/replica.go, line 748 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcKNVr8rgguDhGreR6:-KWcKNVr8rgguDhGreR7:b81bsc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/replica.go#L748)):*
> ```Go
> 		// the preemptive snapshot we'll be in the same state as before the
> 		// preemptive snapshot (old replica present).
> 		return r.setTombstoneKey(ctx, r.store.Engine(), r.mu.state.Desc)
> ```

`setReplicaIDLocked` is too soon for this - this is called as soon as we get the message, before we've decided whether we'll apply the snapshot or not. We might decide not to apply it at all, in which case the old replica is intact. We need to wait to closer to the call to `applySnapshot`.

---

*[pkg/storage/replica.go, line 750 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcLD18FMklh5oTHWMb:-KWcLD18FMklh5oTHWMc:b-j7837t) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/replica.go#L750)):*
> ```Go
> 		return r.setTombstoneKey(ctx, r.store.Engine(), r.mu.state.Desc)
> 	}
> 	if replicaID < r.mu.minReplicaID {
> ```

If we move this below the check for `r.mu.replicaID == replicaID` then it should be rare enough to read from disk on demand and we don't need the separate `minReplicaID` field in memory (which could get out of sync). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-:-KWcKkwls_ddgyFl79-g:boglxxs)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 4 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 744 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcKamm2h3OHG8mCOX4:-KWcVeWcP9Uv-rsWxbu6:b-vj3cd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/replica.go#L744)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I believe persisting the tombstone to disk is necessary. Otherwise we could still get messages for the old replica ID, which would set the replica ID of our replica and cause the same problems.
</details>
The setting of `r.mu.minReplicaID` above will prevent us from receiving messages for the old replica ID. At least, that is the intent.

---

*[pkg/storage/replica.go, line 748 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcKNVr8rgguDhGreR6:-KWcWQb6Md0OlKW950Ts:baadn4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/replica.go#L748)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> `setReplicaIDLocked` is too soon for this - this is called as soon as we get the message, before we've decided whether we'll apply the snapshot or not. We might decide not to apply it at all, in which case the old replica is intact. We need to wait to closer to the call to `applySnapshot`.
</details>
Hmm, let me think about this. The old replica is supposed to be invalid at this point (i.e. it has been removed from the range but not GC'd yet).

---

*[pkg/storage/replica.go, line 750 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcLD18FMklh5oTHWMb:-KWcWKe-ClYRP0IwG2fN:b-cr11h5) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/replica.go#L750)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> If we move this below the check for `r.mu.replicaID == replicaID` then it should be rare enough to read from disk on demand and we don't need the separate `minReplicaID` field in memory (which could get out of sync). 
</details>
I'm mildly anxious about ready from disk here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 1998 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcXFiFBoe7KLT3VyPs:-KWcXFiGC17XgmcEwgks:b-b2ibjs) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/client_raft_test.go#L1998)):*
> ```Go
> 	const rangeID = roachpb.RangeID(1)
> 	// Up-replicate to a bunch of nodes which stresses a condition where a
> 	// replica created via a preemptive snapshot receives a message for previous
> ```

this sentence is mangled

---

*[pkg/storage/replica.go, line 709 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcXYKaJBuPZb4E_nKr:-KWcXYKaJBuPZb4E_nKs:bl1jltp) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/replica.go#L709)):*
> ```Go
> 	// node without having a replica ID of at least desc.NextReplicaID.
> 	ctx := r.AnnotateCtx(context.TODO())
> 	if err := r.setTombstoneKey(ctx, batch, desc); err != nil {
> ```

caller has a context, just sayin'

---

*[pkg/storage/replica.go, line 735 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcXf3CsZdUmAcmepr6:-KWcXf3CsZdUmAcmepr7:b-qespg8) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/replica.go#L735)):*
> ```Go
> func (r *Replica) setReplicaIDLocked(replicaID roachpb.ReplicaID) error {
> 	if replicaID == 0 {
> 		// If the incoming message didn't give us a new replica ID it is a
> ```

s/is/was/ since you're using past tense.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2000 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWdnFJHOA6l79NCH9yz:-KWdnFJHOA6l79NCH9z-:bx1wwwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/97fc14bf52c37749923d14db6809795a4aab029d/pkg/storage/client_raft_test.go#L2000)):*
> ```Go
> 	// replica created via a preemptive snapshot receives a message for a
> 	// previous incarnation of the replica (i.e. has a smaller replica ID) that
> 	// existed on the same node.
> ```

s/node/store/

---

*[pkg/storage/replica.go, line 709 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcXYKaJBuPZb4E_nKr:-KWdnH_vhIZ7tjdcfBQS:byxq2gq) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/replica.go#L709)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `Store.removeReplicaImpl`? It uses `context.TODO()` as well. 
</details>
Not to hold up this PR, but in general, even if the caller has a context.TODO, it's still worth plumbing (since the caller will have already annotated that context in order to use it).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2001 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWabDpPWDJ260ndw09I:-KWfR_Gvp7jGA5Xwew1M:b-7dfbo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/client_raft_test.go#L2001)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Added a comment, though I'm not sure it is the heartbeat traffic. Perhaps the extra replicas cause concurrent down-replication. But I wasn't able to reproduce the problem with 4 replicas.
</details>
Ah, I forgot that the replication queue would be concurrently trying to down-replicate those. You could disable the replication queue to see if that's the cause. I think I'd be more comfortable with the replication queue disabled here; if it is in fact exposing the races there's probably something else we could do to make it more reliable. The heartbeats seem like a more plausible solution to me. 

---

*[pkg/storage/replica.go, line 744 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcKamm2h3OHG8mCOX4:-KWfT260DVRiFbgA16VB:b-5hjxzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/replica.go#L744)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The setting of `r.mu.minReplicaID` above will prevent us from receiving messages for the old replica ID. At least, that is the intent.
</details>
It will prevent us from receiving messages for the old replica ID for the life of this process, but we could still get more after restart (I think the way the raft transport queues things up would allow for this).

---

*[pkg/storage/replica.go, line 748 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcKNVr8rgguDhGreR6:-KWfSE1RB9XqVerI0sjY:b8zjb1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/replica.go#L748)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Hmm, let me think about this. The old replica is supposed to be invalid at this point (i.e. it has been removed from the range but not GC'd yet).
</details>
The old replica is believed to be invalid by the sender of the preemptive snapshot, at the time that it started the process. I think our only guarantee that this is (still) in fact the case is whether raft decides to apply the snapshot or not. 

---

*[pkg/storage/replica.go, line 750 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcLD18FMklh5oTHWMb:-KWfS1mNT4ILVtetv3k-:bsmci3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/replica.go#L750)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I'm mildly anxious about ready from disk here.
</details>
And I'm mildly anxious about the risk of these values diverging :)  Replica IDs change very rarely, so it should be fine. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
#9037 was first found in production, so this doesn't just deflake a test. The first line of the commit message should refer to the production change. (When writing the release notes I ignore changes that just deflake tests, but this deserves to be called out as a bug fix when it goes in).

petermattis commented:
@bdarnell Noted about the commit message. Will adjust.

tamird commented:
Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_raft_test.go, line 2001 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWabDpPWDJ260ndw09I:-KX6HumDig2a2KyPWNhu:bs9tlv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/client_raft_test.go#L2001)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Ah, I forgot that the replication queue would be concurrently trying to down-replicate those. You could disable the replication queue to see if that's the cause. I think I'd be more comfortable with the replication queue disabled here; if it is in fact exposing the races there's probably something else we could do to make it more reliable. The heartbeats seem like a more plausible solution to me. 
</details>
Disabling the replication queue did not make the test more reliable, so your theory about heartbeats sounds more plausible now. 

---

*[pkg/storage/client_raft_test.go, line 2000 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWdnFJHOA6l79NCH9yz:-KX6GT44Jpqp_KcFb0zP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/97fc14bf52c37749923d14db6809795a4aab029d/pkg/storage/client_raft_test.go#L2000)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/node/store/
</details>
Done.

---

*[pkg/storage/replica.go, line 748 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcKNVr8rgguDhGreR6:-KX6OyhxGQonycNUguhA:b-jmlxeu) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/replica.go#L748)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The old replica is believed to be invalid by the sender of the preemptive snapshot, at the time that it started the process. I think our only guarantee that this is (still) in fact the case is whether raft decides to apply the snapshot or not. 
</details>
I've moved setting of the tombstone key to just before the preemptive snapshot is applied.

---

*[pkg/storage/replica.go, line 750 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWcLD18FMklh5oTHWMb:-KX6aBhKsolsvb7PpXlQ:b-c3y9xd) ([raw file](https://github.com/cockroachdb/cockroach/blob/85e6202e55ebcb3e390d83ada7014ec565f5165c/pkg/storage/replica.go#L750)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> And I'm mildly anxious about the risk of these values diverging :)  Replica IDs change very rarely, so it should be fine. 
</details>
True, the replica IDs change rarely. A more practical concern, using `r.mu.minReplicaID` makes testing of the logic here easier.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2001 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWabDpPWDJ260ndw09I:-KX6kcyJuIrWdHsJ4UAr:bkgt489) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/client_raft_test.go#L2001)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Disabling the replication queue did not make the test more reliable, so your theory about heartbeats sounds more plausible now. 
</details>
Might be worth noting in the comment.

---

*[pkg/storage/client_raft_test.go, line 2000 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWdnFJHOA6l79NCH9yz:-KX6kezpSunupl9-CbqE:b-p9poot) ([raw file](https://github.com/cockroachdb/cockroach/blob/97fc14bf52c37749923d14db6809795a4aab029d/pkg/storage/client_raft_test.go#L2000)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Done.
</details>
is it?

---

*[pkg/storage/replica.go, line 340 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KX6koRknKLHaGTpccks:-KX6koRlCc6MFVHqRXNL:b-17q3k4) ([raw file](https://github.com/cockroachdb/cockroach/blob/393b6e1123071e67f827bbc424f714b8d97249ff/pkg/storage/replica.go#L340)):*
> ```Go
> 		// part of a Raft group.
> 		replicaID roachpb.ReplicaID
> 		// The minimum allowed ID for this replica. Initialized from
> ```

this wrapping looks odd

---

*[pkg/storage/replica_test.go, line 6052 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KX6lXhp1tvSl1BXp0px:-KX6lXhp1tvSl1BXp0py:b-hdt2dw) ([raw file](https://github.com/cockroachdb/cockroach/blob/393b6e1123071e67f827bbc424f714b8d97249ff/pkg/storage/replica_test.go#L6052)):*
> ```Go
> 		{0, 1, 1, ""},
> 		{0, 2, 1, "raft group deleted"},
> 		{1, 2, 1, ""},
> ```

why is there no error here? newReplicaID is lower than minReplicaID, isn't it?

---

*[pkg/storage/replica_test.go, line 6065 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KX6lWujLSDbZbGT8CEV:-KX6lWujLSDbZbGT8CEW:b32v1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/393b6e1123071e67f827bbc424f714b8d97249ff/pkg/storage/replica_test.go#L6065)):*
> ```Go
> 			if c.expected == "" {
> 				if err != nil {
> 					t.Fatalf("expected success, but found %v", err)
> ```

%s

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2001 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWabDpPWDJ260ndw09I:-KX6mUvWwZkXEm3463PL:b-k2lo3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/client_raft_test.go#L2001)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Might be worth noting in the comment.
</details>
I'm not clear on what the comment would say. Where is the extra heartbeat traffic to nodeIdx 2 coming from? I'm actually not at all sure why adding these extra replicas makes the test flakier. 

---

*[pkg/storage/client_raft_test.go, line 2000 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWdnFJHOA6l79NCH9yz:-KX6mLCrK_EfCD586FOd:b-ly6j7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/97fc14bf52c37749923d14db6809795a4aab029d/pkg/storage/client_raft_test.go#L2000)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is it?
</details>
Huh. I definitely made that change. Done. Again.

---

*[pkg/storage/replica.go, line 340 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KX6koRknKLHaGTpccks:-KX6n-ZjUS5CWqGMUTzN:ba1e6xs) ([raw file](https://github.com/cockroachdb/cockroach/blob/393b6e1123071e67f827bbc424f714b8d97249ff/pkg/storage/replica.go#L340)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this wrapping looks odd
</details>
Emacs is insisting on it. Unwrapping creates an unwieldy 93-character line.

---

*[pkg/storage/replica_test.go, line 6052 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KX6lXhp1tvSl1BXp0px:-KX6n6VcpXUfaUazzy5n:b-70zlkj) ([raw file](https://github.com/cockroachdb/cockroach/blob/393b6e1123071e67f827bbc424f714b8d97249ff/pkg/storage/replica_test.go#L6052)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why is there no error here? newReplicaID is lower than minReplicaID, isn't it?
</details>
Because we check existing replica ID equal to new replica ID first. This condition should be impossible.

---

*[pkg/storage/replica_test.go, line 6065 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KX6lWujLSDbZbGT8CEV:-KX6nCelc8tpM-JMkwco:b32h8nh) ([raw file](https://github.com/cockroachdb/cockroach/blob/393b6e1123071e67f827bbc424f714b8d97249ff/pkg/storage/replica_test.go#L6065)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> %s
</details>
You bring up this suggestion repeatedly in reviews, but I don't see this as being anything other than your preference. Changing this certainly won't affect this test or affect our code quality.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/storage/client_raft_test.go, line 2001 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KWabDpPWDJ260ndw09I:-KX6pIUoMeQxrcZEWCBk:bg4ag7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f3aea402cca8798c1cf836e7f9fdaae546c77cb/pkg/storage/client_raft_test.go#L2001)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I'm not clear on what the comment would say. Where is the extra heartbeat traffic to nodeIdx 2 coming from? I'm actually not at all sure why adding these extra replicas makes the test flakier. 
</details>
Well that's a bummer.

---

*[pkg/storage/replica.go, line 340 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KX6koRknKLHaGTpccks:-KX6pPH3nddz6RKji7Zz:b5s8xv6) ([raw file](https://github.com/cockroachdb/cockroach/blob/393b6e1123071e67f827bbc424f714b8d97249ff/pkg/storage/replica.go#L340)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Emacs is insisting on it. Unwrapping creates an unwieldy 93-character line.
</details>
Ah, the next word is real big.

---

*[pkg/storage/replica_test.go, line 6052 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KX6lXhp1tvSl1BXp0px:-KX6pX49yIuuEPJJNGor:b-6lo0gg) ([raw file](https://github.com/cockroachdb/cockroach/blob/393b6e1123071e67f827bbc424f714b8d97249ff/pkg/storage/replica_test.go#L6052)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Because we check existing replica ID equal to new replica ID first. This condition should be impossible.
</details>
Ouch. Seems like we should remove this test case or at least have a comment on it.

---

*[pkg/storage/replica_test.go, line 6065 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KX6lWujLSDbZbGT8CEV:-KX6pb03IZiLFcFi2O11:b-pf3ckj) ([raw file](https://github.com/cockroachdb/cockroach/blob/393b6e1123071e67f827bbc424f714b8d97249ff/pkg/storage/replica_test.go#L6065)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> You bring up this suggestion repeatedly in reviews, but I don't see this as being anything other than your preference. Changing this certainly won't affect this test or affect our code quality.
</details>
Like many other stylistic choices, this one is about giving the reader a hint about the expected value of a variable without having the page in the surrounding code.

In general, I try to avoid leaving stylistic comments in the absence of other non-stylistic comments, but when the cost of a CI cycle is already present, the incremental cost of such suggestions approaches zero.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_test.go, line 6052 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KX6lXhp1tvSl1BXp0px:-KX6t6Dn8CzfqKosNVeO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/393b6e1123071e67f827bbc424f714b8d97249ff/pkg/storage/replica_test.go#L6052)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Ouch. Seems like we should remove this test case or at least have a comment on it.
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6065 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690#-KX6lWujLSDbZbGT8CEV:-KX6s2rTsNTQF5LgpOv3:bw3pnc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/393b6e1123071e67f827bbc424f714b8d97249ff/pkg/storage/replica_test.go#L6065)):*
> Like many other stylistic choices, this one is about giving the reader a hint about the expected value of a variable without having the page in the surrounding code.

In my opinion, a reader shouldn't be inferring the nil-ness of a parameter from its formatting variable. That's just asking to be lied to.

> the incremental cost of such suggestions approaches zero.

These suggestions are not zero cost. Filling the review space with such suggestions can give a false sense of security that a change has been reviewed in depth. This one in particular is inconsequential, but still adds a bit of time to address. We've already spent too much time discussing it. You might argue that it would have been quicker if I just changed the code and moved on, which is true locally, but not globally. 

I'm not saying that we should avoid all stylistic comments, but some are more nit-picky that others. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 2 of 4 files at r5, 1 of 1 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10690)*
<!-- Sent from Reviewable.io -->
